pkg/operator/target_status_test.go

```diff
@@ -0,0 +1,1532 @@
+		Status: corev1.PodStatus{
+			Phase: corev1.PodRunning,
+			PodIP: "127.0.0.1",
+			ContainerStatuses: []corev1.ContainerStatus{{
+				Name:  "prometheus",
+				Ready: true,
+			}},
+		},
+	}
+
+	kubeClient := fake.NewClientBuilder().WithScheme(scheme).WithObjects(&appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      NameCollector,
+			Namespace: opts.OperatorNamespace,
+		},
+		Spec: appsv1.DaemonSetSpec{
+			Selector: &metav1.LabelSelector{},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{},
+				Spec: corev1.PodSpec{
+					Containers: []corev1.Container{{
+						Name: "prometheus",
+						Ports: []corev1.ContainerPort{{
+							Name:          "prom-metrics",
+							ContainerPort: port,
+						}},
+					}},
+				},
+			},
+		},
+	}).WithObjects(&monitoringv1.PodMonitoring{
+		ObjectMeta: metav1.ObjectMeta{Name: "prom-example-1", Namespace: "gmp-test"},
+		Spec: v1.PodMonitoringSpec{
+			Endpoints: []v1.ScrapeEndpoint{{
+				Port: intstr.FromString("metrics"),
+			}},
+		},
+	}).WithObjects(pod).Build()
+
+	prometheusTargetMap := make(map[string]*prometheusv1.TargetsResult, 1)
+	key := getPodKey(pod, port)
+	prometheusTargetMap[key] = &prometheusv1.TargetsResult{
+		Active: []prometheusv1.ActiveTarget{{
+			Health: "up",
+			Labels: map[model.LabelName]model.LabelValue{
+				"instance": model.LabelValue("a"),
+			},
+			ScrapePool:         "PodMonitoring/gmp-test/prom-example-1/metrics",
+			LastError:          "err x",
+			LastScrapeDuration: 1.2,
+		}},
+		Dropped: []prometheusv1.DroppedTarget{},
+	}
+
+	targetFetcher := &mockTargetFetcher{
+		prometheusTargets: prometheusTargetMap,
+	}
+
+	ch := make(chan event.GenericEvent, 1)
+	reconciler := &targetStatusReconciler{
+		ch:           ch,
+		opts:         opts,
+		getTarget:    targetFetcher.getTarget,
+		logger:       logger,
+		kubeClient:   kubeClient,
+		timerBuilder: newTestTimerBuilder(timer),
+	}
+
+	expectStatus := func(t *testing.T, description string, expected []monitoringv1.ScrapeEndpointStatus) {
+		// Must poll because status is updated via other thread.
+		err := wait.Poll(100*time.Millisecond, 2*time.Second, func() (done bool, err error) {
+			var podMonitorings monitoringv1.PodMonitoringList
+			kubeClient.List(ctx, &podMonitorings)
+			switch amount := len(podMonitorings.Items); amount {
+			case 0:
+				return false, nil
+			case 1:
+				break
+			default:
+				return false, errors.Errorf("invalid PodMonitorings found: %d", amount)
+			}
+
+			status := podMonitorings.Items[0].Status.EndpointStatuses
+			normalizeEndpointStatuses(status, metav1.Time{})
+			return cmp.Equal(status, expected), nil
+		})
+		if err != nil {
+			if err == wait.ErrWaitTimeout {
+				var podMonitorings monitoringv1.PodMonitoringList
+				kubeClient.List(ctx, &podMonitorings)
+				if len(podMonitorings.Items) == 0 {
+					t.Fatalf("Could not find %s PodMonitoring", description)
+				} else {
+					status := podMonitorings.Items[0].Status.EndpointStatuses
```

Comment:
I don't think this `else` statement could ever happen?

Reply:
It's when the `cmp.Diff` inside the poll fails. Originally, the status is empty so there is a difference but we expect the status to change asynchronously because the reconciler is fired from a goroutine. To make things simpler, I refactored this code to avoid checking inside and outside the poll.